src/analysis/liouville/liouville_constant.lean

```diff
@@ -76,6 +74,116 @@ lemma liouville_number_eq_initial_terms_add_tail (hm : 1 < m) (k : ℕ) :
+... = (1 - 1 / m)⁻¹ * (1 / m ^ (n + 1)!) :
+    -- the series if the geometric series
+    mul_eq_mul_right_iff.mpr (or.inl (tsum_geometric_of_abs_lt_1 mi))
+
+lemma aux_calc (n : ℕ) (hm : 2 ≤ m) :
+  (1 - 1 / m)⁻¹ * (1 / m ^ (n + 1)!) ≤ 1 / (m ^ n!) ^ n :=
+calc (1 - 1 / m)⁻¹ * (1 / m ^ (n + 1)!) ≤ 2 * (1 / m ^ (n + 1)!) :
+  -- the second factors coincide (and are non-negative),
+  -- the first factors, satisfy the inequality `sub_one_div_inv_le_two`
+  mul_mono_nonneg (one_div_nonneg.mpr (pow_nonneg (zero_le_two.trans hm) _))
+    (sub_one_div_inv_le_two hm)
+... = 2 / m ^ (n + 1)! : mul_one_div 2 _
+... = 2 / m ^ (n! * (n + 1)) : congr_arg ((/) 2) (congr_arg (pow m) (mul_comm _ _))
+... ≤ 1 / m ^ (n! * n) :
+  begin
+    -- [ NB: in this block, I do not follow the brace convention for subgoals -- I wait until
+    --   I solve all extraneous goals at once with `exact pow_pos (zero_lt_two.trans_le hm) _`. ]
+    -- Clear denominators and massage*
+    apply (div_le_div_iff _ _).mpr,
+    conv_rhs { rw [one_mul, mul_add, pow_add, mul_one, pow_mul, mul_comm, ← pow_mul] },
+    -- the second factors coincide, so we prove the inequality of the first factors*
+    apply (mul_le_mul_right _).mpr,
+    -- solve all the inequalities `0 < m ^ ??`
+    any_goals { exact pow_pos (zero_lt_two.trans_le hm) _ },
+    -- `2 ≤ m ^ n!` is a consequence of monotonicity of exponentiation at `2 ≤ m`.
+    exact trans (trans hm (pow_one _).symm.le) (pow_mono (one_le_two.trans hm) n.factorial_pos)
+  end
+... = 1 / (m ^ n!) ^ n : congr_arg ((/) 1) (pow_mul m n! n)
+
+end two_useful_inequalities
 
 end m_is_real
+
+section m_is_natural
+variable {m : ℕ}
+
+/--  The sum of the `k` initial terms of the Liouville number to base `m` is a ratio of natural
+numbers where the denominator is `m ^ k!`. -/
+lemma liouville_number_rat_initial_terms (hm : 0 < m) (k : ℕ) :
+∃ p : ℕ, liouville_number_initial_terms m k = p / m ^ k! :=
+begin
+  induction k with k h,
+  { exact ⟨1, by rw [liouville_number_initial_terms, range_one, sum_singleton, nat.cast_one]⟩ },
+  { rcases h with ⟨p_k, h_k⟩,
+    use p_k * (m ^ ((k + 1)! - k!)) + 1,
+    unfold liouville_number_initial_terms at h_k ⊢,
+    rw [sum_range_succ, h_k, div_add_div, div_eq_div_iff, add_mul],
+    { norm_cast,
+      rw [add_mul, one_mul, nat.factorial_succ,
+        show k.succ * k! - k! = (k.succ - 1) * k!, by rw [nat.mul_sub_right_distrib, one_mul],
+        nat.succ_sub_one, nat.succ_eq_add_one, add_mul, one_mul, pow_add],
+      simp [mul_assoc] },
+    refine mul_ne_zero_iff.mpr ⟨_, _⟩,
+    all_goals { exact pow_ne_zero _ (nat.cast_ne_zero.mpr hm.ne.symm) } }
+end
+
+theorem is_liouville (hm : 2 ≤ m) :
+  liouville (liouville_number m) :=
+begin
+  -- two useful inequalities
+  have mZ1 : 1 < (m : ℤ) := nat.cast_one.symm.le.trans_lt
+    (one_lt_two.trans_le (nat.cast_two.symm.le.trans (int.to_nat_le.mp hm))),
+  have m1 : 1 < (m : ℝ) :=
+    one_lt_two.trans_le (nat.cast_two.symm.le.trans (nat.cast_le.mpr hm)),
```

Comment:
```suggestion
  have mZ1 : 1 < (m : ℤ) := by { norm_cast, exact one_lt_two.trans_le hm },
  have m1 : 1 < (m : ℝ) := by { norm_cast, exact one_lt_two.trans_le hm },
```

again, optional but looks nicer imo

Reply:
This is when I was trying to learn what `norm_cast` did and practiced with avoid using it!  Also, in an earlier version of the proof, using the explicit term mode instead of `norm_cast` sped it up quite a bit.  I like your version!